crates/aiken-lang/src/parser/utils.rs

```diff
@@ -0,0 +1,174 @@
+use chumsky::prelude::*;
+
+use crate::ast;
+
+use super::{
+    error::{self, ParseError},
+    token::{Base, Token},
+};
+
+pub fn public() -> impl Parser<Token, (), Error = ParseError> {
+    just(Token::Pub).ignored()
```

Comment:
I think this one could be inline in places it's used to be honest. I find it more confusing to bump into a `utils::public()` than to see its actual definition. One think we could have in `utils` however is the larger pattern that often goes with it: `utils::public().or_not()` and later: `public: public.is_some()`. We also do this for `Token::Opaque`. So there's kind of a pattern for those flags keywords. For example:

```rs
pub fn optional_flag(token: Token) -> impl Parser<Token, bool, Error = ParseError> {
    just(token).ignored().or_not().map(|v| v.is_some())
}
```

which would change, for example:

```rs
    utils::public()
        .then(just(Token::Opaque).ignored().or_not())
        .or_not()
        .then(utils::type_name_with_args())
        .then(choice((constructors, record_sugar)))
        .map_with_span(|((pub_opaque, (name, parameters)), constructors), span| {
            ast::UntypedDefinition::DataType(ast::DataType {
                ...
                opaque: pub_opaque
                    .map(|(_, opt_opaque)| opt_opaque.is_some())
                    .unwrap_or(false),
                public: pub_opaque.is_some(),
            })
        })
```

into:

```rs
    utils::optional_flag(Token::Pub)
        .then(utils::optional_flag(Token::Opaque))
        .then(utils::type_name_with_args())
        .then(choice((constructors, record_sugar)))
        .map_with_span(
            |(((public, opaque), (name, parameters)), constructors), span| {
                ast::UntypedDefinition::DataType(ast::DataType {
                    ...
                    opaque,
                    public,
                })
            },
        )

```

slightly less noise and more reusability IMO.